### PR TITLE
prosemirror-view: make decorations generic [suggestion]

### PR DESCRIPTION
**Pull request by ifiokjr**
I thought it might be a nice addition to make the decoration spec generic. I've found myself working quite closely with decorations lately and this addition would allow workflows like the following. 

```ts
let decorationSet: DecorationSet; // Defined somewhere in the code.

const isSpecialDecoration = (
  decoration: Decoration,
): decoration is Decoration<{ type: 'special'; value: number }> => 
  decoration.spec.type === 'special';

const decorations = decorationSet.find(); // Get all decorations
const filtered = decorations.filter(isSpecialDecoration);

type A = typeof filtered[number]; // => Decoration<{ type: 'special', value: number }>
```

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

### CODE REVIEW

**Comment by ifiokjr** [comment]
Don’t merge yet. I’d like the feedback.

**Review by mrcrane — CHANGES_REQUESTED** [review]
Ok, I will mark this as changes requested for now and I will approve when you get some feedback. Thanks!

**Comment by ifiokjr** [comment]
@timjb @patsimm @davidka @bradleyayers any thoughts on this addition?

I'd rather not just merge it without any thoughts. I've been using it for a few days in my own code and it's made working with decorations more intuitive.

File: types/prosemirror-view/index.d.ts [hunk 1]
```diff
@@ -48,87 +109,32 @@ export class Decoration {
    * also directly pass a DOM node. getPos can be used to find the
    * widget's current document position.
    */
-  static widget(
+  static widget<T extends {} = { [key: string]: any }>(
```
Comment (timjb) [inline types/prosemirror-view/index.d.ts]: Use `static widget<T extends object = { [key: string]: any }>(` instead. The `extends object` makes sure that one gets a type error if for example one tried to use a string as the `spec` argument:

```typescript
Decoration.widget(0, (x) => x.dom, 'this argument should be an object, not a string!'); // $ExpectError
```

(Same for other functions.)

**Review by timjb — CHANGES_REQUESTED** [review]
One small detail, looks good otherwise!

File: types/prosemirror-view/index.d.ts (hunk 1 above)
Comment (ifiokjr) [reply types/prosemirror-view/index.d.ts]: Thanks @timjb! I never knew this. I've just updated several other projects with this enhancement.

**Comment by ifiokjr** [comment]
> One small detail, looks good otherwise!

Thanks @timjb, I've addressed the request. 

@mrcrane I'm happy to merge the changes now.

**Review by timjb — APPROVED** [review]
LGTM!

**Review by mrcrane — APPROVED** [review]

**Pull request merged**